### DISCUSSION TITLE
vs2017编译不能通过

### DISCUSSION THREAD

**Discussion by wyx04143** [CefView/QCefView#109] [General]
vs2017 编译最新代码无法编译通过

**Comment by tishion** [comment]
给出详细错误信息。

**Comment by wyx04143** [comment]
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient.cpp(27): error C2039: “CheckInClient”: 不是“CefViewBrowserApp”的成员
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\include\CefViewBrowserApp.h(20): note: 参见“CefViewBrowserApp”的声明
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient.cpp(30): error C2065: “kCefViewQueryFuntionName”: 未声明的标识符
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient.cpp(31): error C2065: “kCefViewQueryCancelFunctionName”: 未声明的标识符
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient.cpp(37): error C2039: “CheckOutClient”: 不是“CefViewBrowserApp”的成员
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\include\CefViewBrowserApp.h(20): note: 参见“CefViewBrowserApp”的声明
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient.cpp(146): error C2065: “kCefViewClientRenderFocusedNodeChangedMessage”: 未声明的标识符
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient.cpp(148): error C2065: “kCefViewClientRenderInvokeMethodMessage”: 未声明的标识符
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient.cpp(150): error C2065: “kCefViewClientRenderReportJSResultMessage”: 未声明的标识符
4&gt;CefViewBrowserClient_FocusHandler.cpp
4&gt;CefViewBrowserClient_JSDialogHandler.cpp
4&gt;CefViewBrowserClient_KeyboardHandler.cpp
4&gt;CefViewBrowserClient_LifeSpanHandler.cpp
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient_LifeSpanHandler.cpp(76): error C2664: “CefViewQueryHandler::CefViewQueryHandler(const CefViewQueryHandler &amp;)”: 无法将参数 1 从“CefViewBrowserClientDelegateInterface::WeakPtr”转换为“CefViewBrowserDelegateWeakPtr”
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient_LifeSpanHandler.cpp(76): note: 没有可用于执行该转换的用户定义的转换运算符，或者无法调用该运算符
4&gt;CefViewBrowserClient_LoadHandler.cpp
4&gt;CefViewBrowserClient_RenderHandler.cpp
4&gt;CefViewBrowserClient_RequestHandler.cpp
4&gt;CefViewBrowserClient_ResourceRequestHandler.cpp
4&gt;CefViewBrowserHandler.cpp
4&gt;CefViewQueryHandler.cpp
4&gt;CefViewDefaultSchemeHandler.cpp
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_89.0.12+g2b76680+chromium-89.0.4389.90_windows64\include/base/cef_ref_counted.h(308): error C2440: “初始化”: 无法从“T *”转换为“T *”
4&gt;&nbsp; &nbsp; &nbsp; &nbsp; with
4&gt;&nbsp; &nbsp; &nbsp; &nbsp; [
4&gt;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; T=CefCallback
4&gt;&nbsp; &nbsp; &nbsp; &nbsp; ]
4&gt;&nbsp; &nbsp; &nbsp; &nbsp; and
4&gt;&nbsp; &nbsp; &nbsp; &nbsp; [
4&gt;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; T=CefRequestCallback
4&gt;&nbsp; &nbsp; &nbsp; &nbsp; ] (编译源文件 F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient_ResourceRequestHandler.cpp)
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_89.0.12+g2b76680+chromium-89.0.4389.90_windows64\include/base/cef_ref_counted.h(308): note: 与指向的类型无关；强制转换要求 reinterpret_cast、C 样式强制转换或函数样式强制转换 (编译源文件 F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient_ResourceRequestHandler.cpp)
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient_ResourceRequestHandler.cpp(35): note: 参见对正在编译的函数 模板 实例化“scoped_refptr<CefRequestCallback&gt;::scoped_refptr<CefCallback&gt;(const scoped_refptr<CefCallback&gt; &amp;)”的引用
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient_ResourceRequestHandler.cpp(35): note: 参见对正在编译的函数 模板 实例化“scoped_refptr<CefRequestCallback&gt;::scoped_refptr<CefCallback&gt;(const scoped_refptr<CefCallback&gt; &amp;)”的引用
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_89.0.12+g2b76680+chromium-89.0.4389.90_windows64\include/base/cef_ref_counted.h(308): error C2439: “scoped_refptr<CefRequestCallback&gt;::ptr_”: 未能初始化成员 (编译源文件 F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient_ResourceRequestHandler.cpp)
4&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_89.0.12+g2b76680+chromium-89.0.4389.90_windows64\include/base/cef_ref_counted.h(358): note: 参见“scoped_refptr<CefRequestCallback&gt;::ptr_”的声明 (编译源文件 F:\QTSorce\QtCefView\QCefView\CefViewCore\src\CefView\CefBrowserApp\CefViewBrowserClient_ResourceRequestHandler.cpp)
4&gt;已完成生成项目“CefViewCore.vcxproj”的操作 - 失败。
5&gt;------ 已启动全部重新生成: 项目: QCefView, 配置: Debug x64 ------
5&gt;	Automatic MOC for target QCefView
5&gt;Building Custom Rule F:/QTSorce/QtCefView/QCefView/src/CMakeLists.txt
5&gt;mocs_compilation_Debug.cpp
5&gt;QCefConfig.cpp
5&gt;QCefContext.cpp
5&gt;QCefEvent.cpp
5&gt;QCefQuery.cpp
5&gt;QCefSetting.cpp
5&gt;QCefView.cpp
5&gt;CCefAppDelegate.cpp
5&gt;CCefClientDelegate.cpp
5&gt;CCefClientDelegate_ContextMenuHandler.cpp
5&gt;CCefClientDelegate_DisplayHandler.cpp
5&gt;CCefClientDelegate_FocusHandler.cpp
5&gt;CCefClientDelegate_KeyboardHandler.cpp
5&gt;CCefClientDelegate_LifeSpanHandler.cpp
5&gt;CCefClientDelegate_LoadHandler.cpp
5&gt;CCefClientDelegate_RenderHandler.cpp
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\CCefClientDelegate_RenderHandler.cpp(23): error C2440: “初始化”: 无法从“QWindow *”转换为“QWindow”
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\CCefClientDelegate_RenderHandler.cpp(23): note: class“QWindow”的构造函数声明为“explicit”
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\CCefClientDelegate_RenderHandler.cpp(24): error C2819: “QWindow”类型没有重载成员“operator -&gt;”
5&gt;c:\qt\qt5.12.0\5.12.0\msvc2017_64\include\qtwidgets\qwidget.h(70): note: 参见“QWindow”的声明
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\CCefClientDelegate_RenderHandler.cpp(24): note: 是否改用“.”?
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\CCefClientDelegate_RenderHandler.cpp(24): error C2232: “-&gt;QWindow::screen”: 左操作数有“class”类型，使用“.”
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\CCefClientDelegate_RenderHandler.cpp(78): error C2039: “screen”: 不是“QWidget”的成员
5&gt;c:\qt\qt5.12.0\5.12.0\msvc2017_64\include\qtwidgets\qwidget.h(128): note: 参见“QWidget”的声明
5&gt;QCefConfigPrivate.cpp
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\QCefConfigPrivate.cpp(14): error C2065: “kCefViewDefaultUserAgent”: 未声明的标识符
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\QCefConfigPrivate.cpp(19): error C2065: “kCefViewRenderProcessName”: 未声明的标识符
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\QCefConfigPrivate.cpp(22): error C2065: “kCefViewResourceDirectoryName”: 未声明的标识符
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\QCefConfigPrivate.cpp(26): error C2065: “kCefViewLocalesDirectoryName”: 未声明的标识符
5&gt;QCefContextPrivate.cpp
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\QCefContextPrivate.cpp(103): error C2039: “IsSafeToExit”: 不是“CefViewBrowserApp”的成员
5&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\include\CefViewBrowserApp.h(20): note: 参见“CefViewBrowserApp”的声明
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\QCefContextPrivate.cpp(116): error C2039: “IsSafeToExit”: 不是“CefViewBrowserApp”的成员
5&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\include\CefViewBrowserApp.h(20): note: 参见“CefViewBrowserApp”的声明
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\QCefContextPrivate.cpp(116): error C2039: “__this”: 不是“CefViewBrowserApp”的成员
5&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\include\CefViewBrowserApp.h(20): note: 参见“CefViewBrowserApp”的声明
5&gt;QCefEventPrivate.cpp
5&gt;QCefQueryPrivate.cpp
5&gt;正在生成代码...
5&gt;正在编译...
5&gt;QCefSettingPrivate.cpp
5&gt;QCefViewPrivate.cpp
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\QCefViewPrivate.cpp(685): error C2039: “position”: 不是“QWheelEvent”的成员
5&gt;c:\qt\qt5.12.0\5.12.0\msvc2017_64\include\qtwidgets\qwidget.h(73): note: 参见“QWheelEvent”的声明
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\QCefViewPrivate.cpp(698): error C3536: “p”: 初始化之前无法使用
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\QCefViewPrivate.cpp(881): error C2065: “kCefViewClientBrowserTriggerEventMessage”: 未声明的标识符
5&gt;KeyboardUtils.cpp
5&gt;ValueConvertor.cpp
5&gt;F:\QTSorce\QtCefView\QCefView\src\details\utils\ValueConvertor.cpp(43): warning C4267: “参数”: 从“size_t”转换到“int”，可能丢失数据
5&gt;QCefContextPrivate_win.cpp
5&gt;F:\QTSorce\QtCefView\QCefView\src\win\details\QCefContextPrivate_win.cpp(32): error C2661: “CefViewBrowserApp::CefViewBrowserApp”: 没有重载函数接受 2 个参数
5&gt;F:\QTSorce\QtCefView\QCefView\src\win\details\QCefContextPrivate_win.cpp(42): error C3536: “app”: 初始化之前无法使用
5&gt;F:\QTSorce\QtCefView\QCefView\src\win\details\QCefContextPrivate_win.cpp(42): error C2664: “bool CefInitialize(const CefMainArgs &amp;,const CefSettings &amp;,scoped_refptr<CefApp&gt;,void *)”: 无法将参数 3 从“int”转换为“scoped_refptr<CefApp&gt;”
5&gt;F:\QTSorce\QtCefView\QCefView\src\win\details\QCefContextPrivate_win.cpp(42): note: 无构造函数可以接受源类型，或构造函数重载决策不明确
5&gt;F:\QTSorce\QtCefView\QCefView\src\win\details\QCefContextPrivate_win.cpp(47): error C2679: 二进制“=”: 没有找到接受“int”类型的右操作数的运算符(或没有可接受的转换)
5&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_89.0.12+g2b76680+chromium-89.0.4389.90_windows64\include/base/cef_ref_counted.h(340): note: 可能是“scoped_refptr<CefViewBrowserApp&gt; &amp;scoped_refptr<CefViewBrowserApp&gt;::operator =(const scoped_refptr<CefViewBrowserApp&gt; &amp;)”
5&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_89.0.12+g2b76680+chromium-89.0.4389.90_windows64\include/base/cef_ref_counted.h(329): note: 或&nbsp; &nbsp; “scoped_refptr<CefViewBrowserApp&gt; &amp;scoped_refptr<CefViewBrowserApp&gt;::operator =(T *)”
5&gt;&nbsp; &nbsp; &nbsp; &nbsp; with
5&gt;&nbsp; &nbsp; &nbsp; &nbsp; [
5&gt;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; T=CefViewBrowserApp
5&gt;&nbsp; &nbsp; &nbsp; &nbsp; ]
5&gt;F:\QTSorce\QtCefView\QCefView\src\win\details\QCefContextPrivate_win.cpp(47): note: 尝试匹配参数列表“(scoped_refptr<CefViewBrowserApp&gt;, int)”时
5&gt;正在生成代码...
5&gt;已完成生成项目“QCefView.vcxproj”的操作 - 失败。
6&gt;------ 已启动全部重新生成: 项目: ALL_BUILD, 配置: Debug x64 ------
6&gt;Building Custom Rule F:/QTSorce/QtCefView/QCefView/CMakeLists.txt
========== 全部重新生成: 成功 4 个，失败 2 个，跳过 0 个 ==========







------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "CefView/QCefView"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年6月7日(星期二) 上午10:04
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [CefView/QCefView] vs2017编译不能通过 (Discussion #109)





 
给出详细错误信息。

**Comment by tishion** [comment]
CefViewCore没有配置正确，去搜索一下git submodule的用法。

**Comment by wyx04143** [comment]
通过git submodule下载子模块，用cmake3.23.2 生成sln 用vs2017 x64编译还是有很多错误，请指导


F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_95.7.12+g99c4ac0+chromium-95.0.4638.54_windows64\include/base/internal/cef_bind_internal.h(1250): error C2440: “初始化”: 无法从“R (__cdecl *)(base::internal::BindStateBase *,conditional<_Test,const std::string&amp;,const std::string&amp;&gt;::type)”转换为“PolymorphicInvoke”
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; with
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; [
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; R=std::basic_string<char,std::char_traits<char&gt;,std::allocator<char&gt;&gt;
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; ] (编译源文件 F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_95.7.12+g99c4ac0+chromium-95.0.4638.54_windows64\libcef_dll\wrapper\cef_resource_manager.cc)
2&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_95.7.12+g99c4ac0+chromium-95.0.4638.54_windows64\include/base/internal/cef_bind_internal.h(1250): note: 该强制转换要求 reinterpret_cast、C 样式强制转换或函数类型强制转换 (编译源文件 F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_95.7.12+g99c4ac0+chromium-95.0.4638.54_windows64\libcef_dll\wrapper\cef_resource_manager.cc)
2&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_95.7.12+g99c4ac0+chromium-95.0.4638.54_windows64\include/base/cef_bind.h(123): note: 参见对正在编译的函数 模板 实例化“decltype(auto) base::internal::BindImpl<base::RepeatingCallback,std::string(__cdecl &amp;)(const std::string &amp;),&gt;(Functor)”的引用
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; with
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; [
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Functor=std::string (__cdecl &amp;)(const std::string &amp;)
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; ] (编译源文件 F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_95.7.12+g99c4ac0+chromium-95.0.4638.54_windows64\libcef_dll\wrapper\cef_resource_manager.cc)
2&gt;F:\QTSorce\QtCefView\QCefView\CefViewCore\dep\cef_binary_95.7.12+g99c4ac0+chromium-95.0.4638.54_windows64\libcef_dll\wrapper\cef_resource_manager.cc(424): note: 参见对正在编译的函数 模板 实例化“base::RepeatingCallback<std::string (const std::string &amp;)&gt; base::BindRepeating<std::string(__cdecl &amp;)(const std::string &amp;),&gt;(Functor)”的引用
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; with
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; [
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Functor=std::string (__cdecl &amp;)(const std::string &amp;)
2&gt;&nbsp; &nbsp; &nbsp; &nbsp; ]
2&gt;cef_xml_object.cc





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "CefView/QCefView"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年6月26日(星期天) 晚上10:47
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [CefView/QCefView] vs2017编译不能通过 (Discussion #109)





 
CefViewCore没有配置正确，去搜索一下git submodule的用法。

**Comment by tishion** [comment]
这个问题你自己解决一下吧，我这里没有2017的环境，我这里只解决目前项目中使用到的环境工具中的问题。
